### ISSUE TITLE
Session Removed in Anonymous User Conversion

### ISSUE THREAD

**Issue by levigroker** [parse-community/parse-server#5143]
### Issue Description

When attempting to sign up a new user the server responds with an "insufficient auth" error.

This appears to be happening because the user session is removed from the server, presumably resulting in a call to `handleSessionMissingError` which ultimately gives a `insufficient auth` error (somewhat related to #5129 ).

NOTE: We encountered this error after upgrading from `parse_server` 2.7.1 to `parse_server` 2.8.4

### Steps to reproduce

Attempt to "sign up" a user.

Reqest:

		PUT /svcs/api/classes/_User/IeoKj0yfzP HTTP/1.1
		Host: account-test.com
		Content-Type: application/json; charset=utf-8
		Accept: */*
		X-Parse-Session-Token: r:9b76f7bb647c6011c0515cdc9a0a4dbf
		X-Parse-Application-Id: [redacted]
		X-Parse-Client-Key: [redacted]
		X-Parse-Installation-Id: [redacted]
		X-Parse-OS-Version: 12.0 (17G65)
		Accept-Language: en-us
		Accept-Encoding: br, gzip, deflate
		X-Parse-Client-Version: i1.17.1
		Content-Length: 120
		User-Agent: Foo/1 CFNetwork/974.2.1 Darwin/17.7.0
		Connection: keep-alive
		X-Parse-App-Build-Version: 1
		X-Parse-App-Display-Version: 7.5.8

		{"email":"levi+26@test.com","username":"levi+26@test.com","password":"0123456789","authData":{"anonymous":null}}
 
Response:

		HTTP/1.1 400 Bad Request
		Access-Control-Allow-Headers: X-Parse-Master-Key, X-Parse-REST-API-Key, X-Parse-Javascript-Key, X-Parse-Application-Id, X-Parse-Client-Version, X-Parse-Session-Token, X-Requested-With, X-Parse-Revocable-Session, Content-Type
		Access-Control-Allow-Methods: GET,PUT,POST,DELETE,OPTIONS
		Access-Control-Allow-Origin: *
		Access-Control-Expose-Headers: X-Parse-Job-Status-Id, X-Parse-Push-Status-Id
		Content-Type: application/json; charset=utf-8
		Date: Thu, 25 Oct 2018 18:41:11 GMT
		ETag: W/"29-HtV13eqzCrP5h4RxdehNmuclqs4"
		Server: nginx/1.13.12
		X-Powered-By: Express
		Content-Length: 41
		Connection: keep-alive

		{"code":206,"error":"Insufficient auth."}

We can pause execution, query the Parse Dashboard for the `sessionID` and get a session object. Once we resume, and the error is returned, the session is no longer visible on the server.

### Expected Results

The user gets signed up.

### Actual Outcome

"Insufficient auth" error is returned.

### Environment Setup

- **Server**
  - parse-server version : 2.8.4
  - Operating System:     Debian (We are using the Docker `FROM node:8-wheezy` base image)
  - Hardware:             AWS ECS
  - Localhost or remote server?: AWS

- **Database**
  - MongoDB version: 3.2.16
  - Storage engine:  MMAPv1
  - Hardware:        Unknown
  - Localhost or remote server?: ObjectRocket

### Logs/Trace

**Client Info:**

The iOS client is calling

		signUpInBackgroundWithBlock:

Which calls `saveInBackgroundWithBlock`:

            if (self.authData[PFAnonymousUserAuthenticationType] == [NSNull null]) {
                [self saveInBackgroundWithBlock:block];

With this state information: 

		PFUser *pfUser = [PFUser currentUser];
		(lldb) po pfUser
		<PFUser: 0x60000297dd80, objectId: Qi4pz7XSpH, localId: (null)> {
			email = "levi+26@test.com";
			username = "levi+26@test.com";
		}

		(lldb) p [PFAnonymousUtils isLinkedWithUser:pfUser]
		(BOOL) $2 = NO
		(lldb) po self.authData
		{
			anonymous = "<null>";
		}

		(lldb) po pfUser.sessionToken
		r:9b76f7bb647c6011c0515cdc9a0a4dbf

**Server logs**

		2018-10-25T18:41:11.783Z - error: Error generating response. ParseError { code: 206, message: 'Insufficient auth.' } 
		{ error: { code: 206, message: 'Insufficient auth.' } }
		2018-10-25T18:41:11.786Z - info: HTTP PUT /api/classes/_User/Qi4pz7XSpH 
		{ res: 
		   { statusCode: 400,
			 body: { code: 206, error: 'Insufficient auth.' } },
		  req: 
		   { url: '/api/classes/_User/Qi4pz7XSpH',
			 headers: 
			  { host: 'svcs',
				connection: 'close',
				'content-length': '120',
				accept: '*/*',
				'accept-encoding': 'br, gzip, deflate',
				'accept-language': 'en-us',
				'content-type': 'application/json; charset=utf-8',
				'user-agent': 'Foo/1 CFNetwork/974.2.1 Darwin/17.7.0',
				'x-parse-app-build-version': '1',
				'x-parse-app-display-version': '7.5.8',
				'x-parse-application-id': '[redacted]',
				'x-parse-client-key': '[redacted]',
				'x-parse-client-version': 'i1.17.1',
				'x-parse-installation-id': '[redacted]',
				'x-parse-os-version': '12.0 (17G65)',
				'x-parse-session-token': '(hidden)',
				'x-forwarded-for': '67.190.84.164',
				'x-forwarded-port': '443',
				'x-forwarded-proto': 'https' },
			 method: 'PUT',
			 httpVersion: '1.0',
			 originalUrl: '/api/classes/_User/Qi4pz7XSpH',
			 query: {},
			 body: 
			  { email: 'levi+26@test.com',
				username: 'levi+26@test.com',
				password: '(hidden)',
				authData: { anonymous: null } } },
		  responseTime: 228 }

**Comment by flovilmart** [comment]
Can you update to 3.1.0 and reopen the issue if it persists?

**Comment by levigroker** [comment]
I'm sorry, but no. We are not in a position to upgrade to 3 at this time.

**Comment by flovilmart** [comment]
That’s unfortunate as we do not have the bandwidth to support and investigate issues in earlier versions. If you ever came up with a patch, then feel free to open a PR

**Comment by levigroker** [comment]
That's unfortunate since it appears this was introduced in 2.8.1 (I can't get 2.8.0 to build), while 2.7.4 appears fine in my testing. This means the 2.x version is being left in a broken state and any users who are unable/incapable of updating to 3.x will need to use the 2.7.4 version as the "last known good."

**Comment by flovilmart** [comment]
What do you mean incapable of upgrading to 3.x? It’s up to you. If you identify the issue introduced in 2.8.1, feel free to open a PR against the 2.8.x line, I would not mind releasing a patch on this one.

But you’d better jump on the 3.x series

**Comment by gnowland** [comment]
I'm running 3.9.0 and experiencing the same issue

**Comment by 7demo** [comment]
@gnowland 
+1.
I  have the  same question. `Parse error: Insufficient auth. {"code":206,"stack":"Error: Insufficient auth. `

**Comment by gnowland** [comment]
@7demo my issue was resolved when I removed `type="submit"` from my angular `<button>` element, which already had a click event (`(click)="myForm.ngSubmit.emit()"`) bound to it and which was inadvertently causing my `ReactiveForm` to be submitted twice. When I stepped through the stack, call by call, I realized my `.signUp()` call was running twice in quick succession - first time was successful and second time threw the `Parse error: Insufficient auth.` because an Session Token could not be issued for the already authenticated user.

Perhaps something like this is also happening to you? 😇
